lib/server.js

```diff
@@ -10,8 +10,36 @@ module.exports = class HTTPServer extends TCPServer {
 
     super({ allowHalfOpen: false })
 
-    this.on('connection', (socket) => new HTTPServerConnection(this, socket, opts))
+    this._timeout = 0
+
+    this.on('connection', (socket) => {
+      const conn = new HTTPServerConnection(this, socket, opts)
+
+      if (this._timeout) {
+        socket.setTimeout(this._timeout)
+        socket.once('timeout', () => this._ontimeout(conn))
+      }
```

Comment:
We should move this and the `_ontimeout()` method to `HTTPServerConnection` so it's easy to add this missing bit: On upgrade, the `timeout` listener should be removed from the socket.